ratatool-scalacheck/src/main/java/org/apache/beam/sdk/coders/shaded/ScioAvroCoder.java

```diff
@@ -0,0 +1,799 @@
+import org.apache.avro.reflect.ReflectDatumWriter;
+import org.apache.avro.reflect.Union;
+import org.apache.avro.specific.SpecificData;
+import org.apache.avro.specific.SpecificDatumReader;
+import org.apache.avro.specific.SpecificDatumWriter;
+import org.apache.avro.specific.SpecificRecord;
+import org.apache.avro.util.ClassUtils;
+import org.apache.avro.util.Utf8;
+import org.apache.beam.sdk.coders.CannotProvideCoderException;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.CoderProvider;
+import org.apache.beam.sdk.coders.CustomCoder;
+import org.apache.beam.sdk.coders.DefaultCoder;
+import org.apache.beam.sdk.util.EmptyOnDeserializationThreadLocal;
+import org.apache.beam.sdk.values.TypeDescriptor;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Supplier;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Suppliers;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+
+/**
+ *
+ * @todo NOTE: SHADED FROM apache/beam@9375df7. REMOVE WHEN BEAM 2.33.0 IS RELEASED.
+ *
+ * A {@link Coder} using Avro binary format.
+ *
+ * <p>Each instance of {@code AvroCoder<T>} encapsulates an Avro schema for objects of type {@code
+ * T}.
+ *
+ * <p>The Avro schema may be provided explicitly via {@link ScioAvroCoder#of(Class, Schema)} or omitted
+ * via {@link ScioAvroCoder#of(Class)}, in which case it will be inferred using Avro's {@link
+ * org.apache.avro.reflect.ReflectData}.
+ *
+ * <p>For complete details about schema generation and how it can be controlled please see the
+ * {@link org.apache.avro.reflect} package. Only concrete classes with a no-argument constructor can
+ * be mapped to Avro records. All inherited fields that are not static or transient are included.
+ * Fields are not permitted to be null unless annotated by {@link Nullable} or a {@link Union}
+ * schema containing {@code "null"}.
+ *
+ * <p>To use, specify the {@code Coder} type on a PCollection:
+ *
+ * <pre>{@code
+ * PCollection<MyCustomElement> records =
+ *     input.apply(...)
+ *          .setCoder(AvroCoder.of(MyCustomElement.class));
+ * }</pre>
+ *
+ * <p>or annotate the element class using {@code @DefaultCoder}.
+ *
+ * <pre>{@code @DefaultCoder(AvroCoder.class)
+ * public class MyCustomElement {
+ *     ...
+ * }
+ * }</pre>
+ *
+ * <p>The implementation attempts to determine if the Avro encoding of the given type will satisfy
+ * the criteria of {@link Coder#verifyDeterministic} by inspecting both the type and the Schema
+ * provided or generated by Avro. Only coders that are deterministic can be used in {@link
+ * org.apache.beam.sdk.transforms.GroupByKey} operations.
+ *
+ * @param <T> the type of elements handled by this coder
+ */
+@SuppressWarnings({
+    "nullness" // TODO(https://issues.apache.org/jira/browse/BEAM-10402)
+})
+public class ScioAvroCoder<T> extends CustomCoder<T> {
```

Comment:
Copying this over from scio so we don't have to add `scio-core` as a dep for generators.